### DISCUSSION TITLE
Rename Argument using Directive

### DISCUSSION THREAD

**Discussion by skumaravelan** [graphql-java/graphql-java#2291] [Q&A]
Hi all, I am looking for rename directive using graphql Java. Is there a directive available that I can use?

**Comment by knotenpunkt** [comment]
do you mean something like that?

> Customising PropertyDataFetcher
> As mentioned above graphql.schema.PropertyDataFetcher is the default data fetcher for fields in graphql-java and it will use standard patterns for fetching object field values.
> 
> It supports a POJO approach and a Map approach in a Java idiomatic way. By default it assumes that for a graphql field fieldX it can find a POJO property called fieldX or a map key called fieldX if the backing object is a Map.
> 
> However you may have small differences between your graphql schema naming and runtime object naming. For example imagine that Product.description is actually represented as getDesc() in the runtime backing Java object.
> 
> If you are using SDL to specify your schema then you can use the @fetch directive to indicate this remapping.


```
  directive @fetch(from : String!) on FIELD_DEFINITION

    type Product {
        id : ID
        name : String
        description : String @fetch(from:"desc")
        cost : Float
        tax : Float
    }
```

the logic behind @fetch is still implemented in graphql-java, but you must "activate" it/make-it-accessable by writing the directive statement explicit into your graphql-schema like in that code-example here, see line 1

take a look here: https://www.graphql-java.com/documentation/v16/data-fetching/

**Reply by skumaravelan** [reply]
Hi 
I have added  @fetch directive to indicate required remapping, in my resolver, i am using dataFetchingEnvironment.getArguments(); to process the args  using Java map. is there anyway to get the replaced field names using this method. I am not having any direct POJO. My code uses dataFetchingEnvironment.getArguments() to get the field name for back end mapping and further processing.

**Reply by knotenpunkt** [reply]
oh ok remapping input-arguments i don't know if there @fetch will work or if there exists still some 'other' programm-code solving this

maybe someone from the graphql team have an answer for that,

if there exists no such code in the graphql-framework then you can write your own to remap this by writing a  GraphqlDirective which implements SchemaDirectiveWiring and registering it on your graphql instance for example.

**Reply by skumaravelan** [reply]
I don’t have an datafetchingEnvironment on SchemaDirectiveWiring at OnArgument callback method.I want to change only internally, client snd server I have different name mapping. Any pointers and example would help me.

**Reply by knotenpunkt** [reply]
@skumaravelan there is a other solution, but its not so easy:

take a look here: https://github.com/graphql-java/graphql-java/discussions/2189

its the graphql.schema.transform.FieldVisibilitySchemaTransformation

with that you can do all what you wish^^

But you have to play a little bit with it to understand what method on which object do what exactly

@bbakerman linked there also to the test-file where you can see a few examples.

Anyway, the most structures in graphql-java are inmutable, so my idea is the following with SchemarDirectiveWiring or with the SchemaTransformation (with that you can do a lot of more things), to read your directives on the arguments and then don't change the origin (because the origin is inmutable) you can copy the argument to a tmp map and change there the keys to the names, you get from your directive.

 now you have an datafetcher interceptor (its only a secound one^^) and in that you read from the datafetcherenvironment what arguments are passed into it, then you do what i described above, coping, manipulating to a tmp map, and these map you put into the context-object you also get from the datafetchingenvironment.
 
 That datafetcher-interceptor you must create in the SchemaTransformation and after doing that with manipulating and putting to the context you call the origin datafetcher; thats it^^

**Reply by knotenpunkt** [reply]
and in the origin-datafetchers you fetch the data then not from dfe.getArguments but from dfe.getContext().get("myNewArguments") or something like that.

if you want also the propertyFetcher work with that, you must write for that proxy-classes

But as i sayed maybe there is still a better and easier solution @bbakerman can give

**Reply by knotenpunkt** [reply]
@skumaravelan im not sure if you saw this replies here, because they are hidden on default, you must click a "button" to see that, maybe that helps

**Reply by skumaravelan** [reply]
@knotenpunkt : I am trying the follow the code example that is referenced in here. I am not quite following on the datafetcher-interceptor. Do I need to use both SchemarDirectiveWiring, SchemaTransformation? Please give me the steps in the form of code  would help. I am not getting where we are setting up dfe.getContext().get("myNewArguments")  in order for me to retrieve the modified argument.

**Reply by skumaravelan** [reply]
@knotenpunkt  : Could you please help me realizing the solution? i am having challenge in copy the argument to a tmp map and change there the keys to the names, you get from your directive.  And where could I create a context in this algorithm ?  I believe the idea is cool and will work for me, but I am not able to follow the process  like datafetcher-interceptor you must create in the SchemaTransformation  and so on.

**Reply by knotenpunkt** [reply]
@skumaravelan see my answer in a new thread

**Comment by skumaravelan** [comment]
Rename works onField of SchemaDirectiveWiring and has Datafetcher. But I need an example, how i could use the datafetcher for onArgument method. So that the client doesn't' need to know about the field name that app can use internally, Please share an example or show me how to get that working like onField method. When I call, dataFetchingEnvironment.getArguments(), need the field with replaced field name and the values. 

    public GraphQLFieldDefinition onField(SchemaDirectiveWiringEnvironment<GraphQLFieldDefinition> environment) {
        GraphQLFieldDefinition field = (GraphQLFieldDefinition)environment.getElement();
        String fetchName = this.atRenameSupport(field.getName(), field.getDirectives());
        DataFetcher dataFetcher = new PropertyDataFetcher(fetchName);
        return field.transform((builder) -> {
            builder.dataFetcher(dataFetcher);
        });
    }



    @Override
    public GraphQLArgument onArgument(SchemaDirectiveWiringEnvironment<GraphQLArgument> environment) {
        GraphQLArgument graphQLArgument = environment.getElement();
        String fetchName = this.atRenameSupport(graphQLArgument.getName(), graphQLArgument.getDirectives());
        DataFetcher dataFetcher = new PropertyDataFetcher(fetchName);
        return graphQLArgument.transform((builder -> {
            builder.name(fetchName);
        }));
    }

**Reply by knotenpunkt** [reply]
ehm i tried it with onArgument, but that method is nerver called but i added a few argumentsDirectives into my schema,

maybe thats an bug? @bbakerman maybe you can take a look on it if the onArgument-method is called.

i did the following:

`extend type Query{
    xxxxx(yyyyy: [String!] @fetch2(from: "hicks"), ......`
    
 ` @Override
    public GraphQLArgument onArgument(SchemaDirectiveWiringEnvironment<GraphQLArgument> environment) {
        System.out.println("debug-pointx");
.....`  

but its never called

**Reply by skumaravelan** [reply]
I am able to add a  directive in  the fields level and onArgument is getting called, but when I rename the field it is getting changed in the published schema, rather I want to change only internally so that I can see updated values using the DataFetcher dataFetcher. Please help me to resolve, How I can update and read for the internal use.

**Reply by knotenpunkt** [reply]
@skumaravelan take a look on the other answer-thread here in this topic, my replies could be hidden, you must unfold then you can see.... there i described you briefly how you could use internaly other namings

**Reply by skumaravelan** [reply]
I am looking for an example that was described to to use graphql.schema.transform.FieldVisibilitySchemaTransformation And also context based example.  Please share an example that I can try to check out that feature.

**Comment by knotenpunkt** [comment]
```java
public class GraphQLTypeVisitorStubWithCodeRegistry extends GraphQLTypeVisitorStub
{
    @Getter protected GraphQLCodeRegistry.Builder graphQLCodeRegistryBuilder;
    @Getter protected GraphQLSchema graphQLSchema;

    public void registerCodeRegistryBuilder(GraphQLCodeRegistry.Builder graphQLCodeRegistryBuilder)
    {
        this.graphQLCodeRegistryBuilder = graphQLCodeRegistryBuilder;
    }

    public void registerGraphQLSchema(GraphQLSchema graphQLSchema)
    {

        this.graphQLSchema = graphQLSchema;
    }
}
```

```java
public class ArgumentsDirectiveVisitor extends GraphQLTypeVisitorStubWithCodeRegistry
{

 

    @Override
    public TraversalControl visitGraphQLFieldDefinition(GraphQLFieldDefinition fieldDefinition, TraverserContext<GraphQLSchemaElement> context)
    {

    


 GraphQLFieldsContainer parentContainerType = (GraphQLFieldsContainer) context.getParentContext().thisNode();

          newDataFetcher = (dfe) ->{
            final Map<String, Object> originArguments = dfe.getArguments();
            final GraphQLContext dfeContext = (GraphQLContext) dfe.getContext();
            final LinkedHashMap<String, Object> mySpecialArguments = dfeContext.getOrDefault("mySpecialArguments", new LinkedHashMap<>());
            mySpecialArguments.put("...","..."); //do something with the originArguments related to the directives
            return this.graphQLCodeRegistryBuilder.getDataFetcher(parentContainerType, fieldDefinition).get(dfe);

        };

   this.graphQLCodeRegistryBuilder.dataFetcher(coordinates(parentContainerType, fieldDefintion), newDataFetcher);



}
```


So its some code i use changed a little bit to give you an idea, so briefly im not sure which exactly method you must override of GraphQLTypeVisitorStub, but it should show you how you can put into it a datafetcher interceptor, the important code is the newDataetcher-lambda and how you can register it in the registry, for that you must inject the registry.....

because i use different visitors in my code i putted it into a abstract parent class. The registry you get nearly on that location where you boostrap up your graphql engine, i think you will find it...... ok moment i give you some code:

```java
  SchemaTransformer schemaTransformer = new SchemaTransformer();


        GraphQLCodeRegistry.Builder codeRegistryBuilder = newCodeRegistry(graphQLSchema.getCodeRegistry());


        for (GraphQLTypeVisitorStubWithCodeRegistry value : this.graphQLTypeVisitor)
        {
            value.registerCodeRegistryBuilder(codeRegistryBuilder);
            value.registerGraphQLSchema(graphQLSchema);
            graphQLSchema = schemaTransformer.transform(graphQLSchema, value);
        }

        graphQLSchema = graphQLSchema.transform(builder ->
        {
            builder.codeRegistry(codeRegistryBuilder.build());
        });


        this.graphQL = GraphQL.newGraphQL(graphQLSchema).build();
        return this;
```
 
 the graphQLSchema itself i need for something other in my code, that you dont need

**Reply by skumaravelan** [reply]
@knotenpunkt : Thank you so much for sharing the example, I am trying this out to get this working, I may need to  override of GraphQLTypeVisitorStub on  public TraversalControl visitGraphQLArgument(
I have made the following changes in this code.   What is the return statement of visitGraphQLFieldDefinition? When I add return TraversalControl.CONTINUE; am getting stackoverlfow error. 

```
public class ArgumentsDirectiveVisitor extends GraphQLTypeVisitorStubWithCodeRegistry
{


    @Override
    public TraversalControl visitGraphQLFieldDefinition(GraphQLFieldDefinition fieldDefinition, TraverserContext<GraphQLSchemaElement> context) {

        GraphQLFieldsContainer parentContainerType = (GraphQLFieldsContainer) context.getParentContext().thisNode();

        DataFetcher newDataFetcher = (dfe) -> {
            final Map<String, Object> originArguments = dfe.getArguments();
            final GraphQLContext dfeContext = (GraphQLContext) dfe.getContext();
            String fetchName = atRenameSupport(fieldDefinition.getName(), fieldDefinition.getDirectives());
            final LinkedHashMap<String, Object> mySpecialArguments = dfeContext.getOrDefault("mySpecialArguments", new LinkedHashMap<>());
            mySpecialArguments.put(fetchName, fieldDefinition.getName() ); //do something with the originArguments related to the directives
            return this.graphQLCodeRegistryBuilder.getDataFetcher(parentContainerType, fieldDefinition).get(dfe);

        };
        this.graphQLCodeRegistryBuilder.dataFetcher(coordinates(parentContainerType, fieldDefinition), newDataFetcher);
        return TraversalControl.CONTINUE;
    }
    private String atRenameSupport(String fieldName, List<GraphQLDirective> directives) {
        Optional<GraphQLArgument> from = DirectivesUtil.directiveWithArg(directives, GraphQLDirectiveSchemaDirectiveWiring.RENAME, "attribute");
        return (String)from.map((arg) -> {
            return String.valueOf(arg.getValue());
        }).orElse(fieldName);
    }
```

```
        SchemaTransformer schemaTransformer = new SchemaTransformer();
        GraphQLCodeRegistry.Builder codeRegistryBuilder = newCodeRegistry(graphQLSchema.getCodeRegistry());
        GraphQLTypeVisitorStubWithCodeRegistry argumentsDirectiveVisitor =  new ArgumentsDirectiveVisitor();

        graphQLSchema = schemaTransformer.transform(graphQLSchema, argumentsDirectiveVisitor);
        argumentsDirectiveVisitor.registerCodeRegistryBuilder(codeRegistryBuilder);
        argumentsDirectiveVisitor.registerGraphQLSchema(graphQLSchema);
        graphQLSchema = graphQLSchema.transform(builder ->
        {
            builder.codeRegistry(codeRegistryBuilder.build());
        });
        return graphQLSchema;
```

**Reply by skumaravelan** [reply]
@knotenpunkt with the above update, I am getting stack overflow error in ArgumentsDirectiveVisitor. Let me know, how method return is handled in here.

**Reply by skumaravelan** [reply]
@knotenpunkt  Also I am not finding a way to add the context. In the code, we are creating it and it not getting added in the context. 
           final LinkedHashMap<String, Object> mySpecialArguments = dfeContext.getOrDefault("mySpecialArguments", new LinkedHashMap<>());
            mySpecialArguments.put("...","...");

**Reply by knotenpunkt** [reply]
hmmm yeah return TraversalControl.CONTINUE; should be the correct one.
the visitGraphQLFieldDefinition(,,,) throws that stackoverflow? or something other in that method?
i didnt test my/the code(-changes), but hmm i cant see what could be trigger that endless-recursion, maybe you can find it out, where the problem exactly is there.

graphQL has a default context-object, its some kind of concurrent-map if i know that correctly, its GraphQLContext
if you dont override it at the beginning of your execution-input code, then there should be used that default object.

into this default map-object you put another one, the mySpecialArgument map, if that not exists, and then you can fill it by merging or fill it once, so was my idea^^

maybe you must there use also another concurrent map, because of concurrency, ah and i see another problem, maybe you have to use there some kind of namespace for each object in that, because different graphql-endpoints/fetcher can be called in one request and they all have then also only one context-object

and then you put into your mySpecialArguments.put(namespace+"_"+directive-renaming-name, dfe.getArguments("graphQL-Schema-name"))

**Reply by knotenpunkt** [reply]
or if you dont want to use namespaces you can also override the datafetchingenvironment to hold that data, maybe by returning not the origin dfe.getArguments but your custumized, maybe thats even a nicer solution

return this.graphQLCodeRegistryBuilder.getDataFetcher(parentContainerType, fieldDefinition).get(new MyDfeManipulator(dfe));

while MyDfeManipulator is a class extends DataFetchingEnvironment and override only getArguments()

**Reply by skumaravelan** [reply]
@knotenpunkt I have to read my directive in onArgument of SchemaDirectiveWiring or TraversalControl visitGraphQLArgument of SchemaTransformer. DataFetchingEnvironment is not playing nice. We do not have much handle on that that one. 

Is there any example that you came across to rename and manipulate the dfe?  All these manipulation is failing or causing an issue one or another for argument . And also no  GraphQLFieldsContainer for the argument, facing the issue there as well. My directives are listed in argument level in the schema, i am having problem using visitGraphQLFieldDefinition or OnField for this requirement. 
 
My query is defined as search (folder: folderSerach @rename(attribute: "directory"):Folder.

I am hoping to rewrite the argument using with the help of directive. DFE is not getting updated .

**Reply by knotenpunkt** [reply]
you should not manipulated the origin dfe but simulate a manipulation with the help of, i think proxy-pattern is the correct word for it, with a proxy and manipulate class on runtime. 

```java
MyDfeManipulator is a class extends DataFetchingEnvironment
{
  this.origin;
   publicMyDfeManipulator (DataFetchingEnvironment origin)
{
this.origin=origin
}

@Override
public ???? getArguments();
{
take this.origin.getArguments()
change in a tmp-copy the argument-keys
return the tmp-copy

}
}
```

**Reply by skumaravelan** [reply]
@knotenpunkt DataFetchingEnvironment is an interface, idea is to override the getArgument, where will be attaching the new DFE, in  the transformer or schema directiveWiring ?

**Reply by knotenpunkt** [reply]
ah ok if it is an interface than write a special method for getArgument and delegate the other methods to the origin unprocessed.

if directiveWiring works, then you can do it there, the transformer should work in any case (i hope^^)

**Reply by skumaravelan** [reply]
@knotenpunkt Could you please point me, where is the current implementation of getArguments method implementation, I am still searching where in the code the arguments map is getting populated?

**Reply by skumaravelan** [reply]
@knotenpunkt, changing and fetching in through the transformer is quite cumbersome, I am trying to get both list of arguments having a rename directive together in a map form, so that I can apply the key rename using the resultant map. 
How could we combine these  two steps into one? is there any alternative one can suggest ?

            // all the arguments with our @rename directive
            List<String> argsWithLengthDirective = dataFetchingEnvironment.getFieldDefinition()
                    .getArguments()
                    .stream()
                    .filter(arg -> arg.getDirective(RENAME) != null)
                    .map(GraphQLArgument::getName)
                    .collect(Collectors.toList());

            for (String argName : argsWithLengthDirective) {
                GraphQLArgument argument = dataFetchingEnvironment.getFieldDefinition().getArgument(argName);
                GraphQLDirective renameDirective = argument.getDirective("rename");
                newMap.put(argument.getName(), String.valueOf(renameDirective.getArgument("attribute").getValue()));
            }
